app/code/AstoundDRudenko/PriceBadge/Pricing/Price/Previous/LowestPriceOptionsProvider.php

```diff
@@ -0,0 +1,83 @@
+<?php
+
+declare(strict_types=1);
+namespace AstoundDRudenko\PriceBadge\Pricing\Price\Previous;
+
+use Magento\Catalog\Model\ResourceModel\Product\CollectionFactory;
+use Magento\Catalog\Model\ResourceModel\Product\LinkedProductSelectBuilderInterface;
+use \Magento\ConfigurableProduct\Pricing\Price\LowestPriceOptionsProviderInterface;
+use Magento\Catalog\Api\Data\ProductInterface;
+use Magento\Framework\App\ResourceConnection;
+use Magento\Store\Model\StoreManagerInterface;
+
+/**
+ * Retrieve list of products where each product contains lower price than others at least for one possible price type
+ */
+class LowestPriceOptionsProvider implements LowestPriceOptionsProviderInterface
+{
+    /**
+     * @var ResourceConnection
+     */
+    private $resource;
+
+    /**
+     * @var LinkedProductSelectBuilderInterface
+     */
+    private $linkedProductSelectBuilder;
+
+    /**
+     * @var CollectionFactory
+     */
+    private $collectionFactory;
+
+    /**
+     * @var StoreManagerInterface
+     */
+    private $storeManager;
+
+    /**
+     * Key is product id and store id. Value is array of prepared linked products
+     *
+     * @var array
+     */
+    private $linkedProductMap;
+
+    /**
+     * @param ResourceConnection $resourceConnection
+     * @param LinkedProductSelectBuilderInterface $linkedProductSelectBuilder
+     * @param CollectionFactory $collectionFactory
+     * @param StoreManagerInterface $storeManager
+     */
+    public function __construct(
+        ResourceConnection $resourceConnection,
+        LinkedProductSelectBuilderInterface $linkedProductSelectBuilder,
+        CollectionFactory $collectionFactory,
+        StoreManagerInterface $storeManager
+    ) {
+        $this->resource = $resourceConnection;
+        $this->linkedProductSelectBuilder = $linkedProductSelectBuilder;
+        $this->collectionFactory = $collectionFactory;
+        $this->storeManager = $storeManager;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getProducts(ProductInterface $product)
```

Comment:
when you made some changes to a method - override it - please comment as much as possible.
e.g.:
```
/**
    * Override. Added 'previous_price' to the result.
     * See 'AstoundDRundeko change(s)' comment.
     * {@inheritdoc}
     */
    public function getProducts(ProductInterface $product)
     {
        $key = $this->storeManager->getStore()->getId() . '-' . $product->getId();
        if (!isset($this->linkedProductMap[$key])) {
            $productIds = $this->resource->getConnection()->fetchCol(
                '(' . implode(') UNION (', $this->linkedProductSelectBuilder->build($product->getId())) . ')'
            );
            $this->linkedProductMap[$key] = $this->collectionFactory->create()
                ->addAttributeToSelect(
                    ['price', 'special_price', 'special_from_date', 'special_to_date', 'tax_class_id', 'previous_price']//AstoundDRundeko change
                )
                ->addIdFilter($productIds)
                ->getItems();
        }
       //AstoundDRundeko changes region start
           //this is an example of comments for some multiline changes
       //endregion
        return $this->linkedProductMap[$key];
    }
```